### ISSUE TITLE
Forward to wrong container if DEFAULT_HOST has not HTTPS, other has and request is HTTPS.

### ISSUE THREAD

**Issue by iBobik** [nginx-proxy/nginx-proxy#698]
I have stack like this:
```
proxy:
  build: ./proxy
  ports:
    - "80:80"
    - "443:443"
  volumes:
    - /var/run/docker.sock:/tmp/docker.sock:ro
    - nginx_certs:/etc/nginx/certs           # There is certificate for example.com only
  environment:
    DEFAULT_HOST: static-hosting

static:
  build: ./static
  environment:
    VIRTUAL_HOST: static-hosting

app:
  build: ./app
  environment:
    VIRTUAL_HOST: example.com
```

It generates something like this (simplified to contain only important parts):
```
# static-hosting
server {
	server_name static-hosting;
	listen 80 default_server;
}
# example.com
server {
	server_name example.com;
	listen 80 ;
	return 301 https://$host$request_uri;
}
server {
	server_name example.com;
	listen 443 ssl http2 ;
}
```

This config behaves like this:
- http://example.com → app
- http://static-example.com → static
- https://example.com → app
- https://static-example.com → **app but should static** (also with certificate for app)

Why? Because Nginx see default_server only for port 80. For 443 it does not see default_server, so it use the first server block (what has 443 implemented).

More info about request processing: http://nginx.org/en/docs/http/request_processing.html

Fix can be to add 443 server even for http services (so Nginx will catch response) and do something like error or try to redirect (if it is possible without certificate).

**Comment by iBobik** [comment]
> special nginx’s non-standard code 444 is returned that closes the connection
So maybe to add something like this for http-only services:
```
server {
        server_name static-hosting;
	listen 443 ssl http2 default_server;
        return      444;
}
```

**Comment by thomasleveil** [comment]
Do you have a server certificate and key files for the static-hosting vhost?

**Comment by iBobik** [comment]
No. If I had, this will not happen.

Jan Pobořil
http://honza.poboril.cz/

28. 1. 2017 v 9:12, Thomas LÉVEIL <notifications@github.com>:

**Comment by bjoernbusch** [comment]
Same issue as in #614

**Comment by buchdag** [comment]
This was fixed in #2186
